samplers/aws/xray/remote_sampler.go

```diff
@@ -0,0 +1,186 @@
+// limitations under the License.
+
+package xray // import "go.opentelemetry.io/contrib/samplers/aws/xray"
+
+import (
+	"context"
+	"time"
+
+	"go.opentelemetry.io/contrib/samplers/aws/xray/internal"
+	sdktrace "go.opentelemetry.io/otel/sdk/trace"
+
+	"github.com/go-logr/logr"
+)
+
+// remoteSampler is a sampler for AWS X-Ray which polls sampling rules and sampling targets
+// to make a sampling decision based on rules set by users on AWS X-Ray console
+type remoteSampler struct {
+	// manifest is the list of known centralized sampling rules.
+	manifest *internal.Manifest
+
+	// pollerStarted, if true represents rule and target pollers are started.
+	pollerStarted bool
+
+	// samplingRulesPollingInterval, default is 300 seconds.
+	samplingRulesPollingInterval time.Duration
+
+	// matching attribute
+	serviceName string
+
+	// matching attribute
+	cloudPlatform string
+
+	// fallback sampler
+	fallbackSampler *FallbackSampler
+
+	// logger for logging
+	logger logr.Logger
+}
+
+// Compile time assertion that remoteSampler implements the Sampler interface.
+var _ sdktrace.Sampler = (*remoteSampler)(nil)
+
+// NewRemoteSampler returns a sampler which decides to sample a given request or not
+// based on the sampling rules set by users on AWS X-Ray console. Sampler also periodically polls
+// sampling rules and sampling targets.
+func NewRemoteSampler(ctx context.Context, serviceName string, cloudPlatform string, opts ...Option) (sdktrace.Sampler, error) {
+	// create new config based on options or set to default values
+	cfg, err := newConfig(opts...)
+	if err != nil {
+		return nil, err
+	}
+
+	// create manifest with config
+	m, err := internal.NewManifest(cfg.endpoint, cfg.logger)
+	if err != nil {
+		return nil, err
+	}
+
+	remoteSampler := &remoteSampler{
+		manifest:                     m,
+		samplingRulesPollingInterval: cfg.samplingRulesPollingInterval,
+		fallbackSampler:              NewFallbackSampler(),
+		serviceName:                  serviceName,
+		cloudPlatform:                cloudPlatform,
+		logger:                       cfg.logger,
+	}
+
+	// starts the rule and target poller
+	remoteSampler.start(ctx)
+
+	return remoteSampler, nil
+}
+
+// ShouldSample matches span attributes with retrieved sampling rules and perform sampling,
+// if rules does not match or manifest is expired then use fallback sampling.
+func (rs *remoteSampler) ShouldSample(parameters sdktrace.SamplingParameters) sdktrace.SamplingResult {
+	if rs.manifest.Expired() {
+		// Use fallback sampler if manifest is expired
+		rs.logger.V(5).Info("manifest is expired so using fallback sampling strategy")
+
+		return rs.fallbackSampler.ShouldSample(parameters)
+	}
+
+	// match against known rules
+	r, match, err := rs.manifest.MatchAgainstManifestRules(parameters, rs.serviceName, rs.cloudPlatform)
+	if err != nil {
+		rs.logger.Error(err, "regexp matching error while matching span and resource attributes")
+		return sdktrace.SamplingResult{}
```

Comment:
I don't think so because fallback sampler should be used in 2 cases.

1. when manifest is expired
2. when there is no rule match

This case indicates that we're not able to do the rule match due to some error so we should return the error rather then continue with sampling decision.

Reply:
Sure, but you are returning a sampling result. That result will be used as is, the caller has no concept of the error you encountered and is going to drop the span based on the returned value. That does not seem ideal, especially since we have something more intelligent to fallback to.

Reply:
Yeah I'm not sure if we should use fallback sampler here? I feel that is kind of misrepresentation of the case when there is an error in matching logic.  `return sdktrace.SamplingResult{}` would return SamplingDecision as `DROP`?